samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -127,6 +129,14 @@ public String toString() {
     }
   }
 
+  /**
+   * Supplier interface to allow creation of {@link ProcessorLifecycleListener} for a {@link StreamProcessor}
+   */
+  public interface StreamProcessorListenerSupplier extends Function<StreamProcessor, ProcessorLifecycleListener> {
```

Comment:
The only reason I add it here is to avoid the nuance that you have to create the listener and re-associate the StreamProcessor instance w/ it afterward. If it is too confusing, I can revert back to the old code that requires creation of listener and processor separately and associate them after the creation.

Reply:
Yeah, makes sense. I don't like the create-then-set logic either.
It's a mouthful, but maybe call it StreamProcessorLifecycleListenerFactory for consistency with ProcessorLifecycleListenerFactory? That way the interface and usage is similar and should be less confusing. 

Minor: As with other inner classes and interfaces, prefer moving to the bottom, or better, extracting as a separate class.